### PR TITLE
New api proposal (review in depth requested)

### PR DESCRIPTION
**Pull request by dshadowwolf**
Adds a new annotation based API
Updates to version 1.1 of the json format
Adds biome blacklisting and biome dictionary support
Stops the code from creating the OS1 config directory if it doesn't exist
Preparations in place for the next run of updates

### CODE REVIEW

File: src/main/java/com/mcmoddev/orespawn/impl/os3/OS3APIImpl.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import com.mcmoddev.orespawn.api.os3.OS3API;
+import com.mcmoddev.orespawn.data.ReplacementsRegistry;
+import com.mcmoddev.orespawn.worldgen.OreSpawnWorldGen;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.state.IBlockState;
+import net.minecraftforge.fml.common.registry.GameRegistry;
+
+public class OS3APIImpl implements OS3API {
+	private final Map<String, BuilderLogic> logic;
```
Comment (pau101) [inline src/main/java/com/mcmoddev/orespawn/impl/os3/OS3APIImpl.java:25]: My builder suggestion was to only hold immutable objects which contain the data for generation. Nothing has been done to implement that besides renaming the classes.

File: src/main/java/com/mcmoddev/orespawn/api/os3/BuilderLogic.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+import javax.annotation.Nonnull;
+
+import com.google.common.collect.ImmutableMap;
+
+public interface BuilderLogic {
+	// at this level all we track are the dimensions, really
+	DimensionBuilder DimensionBuilder( @Nonnull String name ); // use "+" to mean "all overworld"
+	DimensionBuilder DimensionBuilder( @Nonnull int id ); // legacy support - referencing dimensions this way needs to stop
+	DimensionBuilder DimensionBuilder(); // literally all dimensions
```
Comment (pau101) [inline src/main/java/com/mcmoddev/orespawn/api/os3/BuilderLogic.java]: I would not have uppercase methods names. `newDimensionBuilder` would be conventional. Also you have @​Nonnull on primitive.

File: src/main/java/com/mcmoddev/orespawn/api/os3/DimensionBuilder.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+import java.util.Optional;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import com.google.common.collect.ImmutableList;
+
+public interface DimensionBuilder {
+  SpawnBuilder SpawnBuilder( @Nullable String name );
```
Comment (pau101) [inline src/main/java/com/mcmoddev/orespawn/api/os3/DimensionBuilder.java]: Uppercase method name, `newSpawnBuilder` would be conventional.

File: src/main/java/com/mcmoddev/orespawn/api/os3/SpawnBuilder.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+import com.google.common.collect.ImmutableList;
+import com.mcmoddev.orespawn.api.BiomeLocation;
+
+import net.minecraft.block.state.IBlockState;
+
+public interface SpawnBuilder {
+	FeatureBuilder FeatureBuilder( @Nullable String featureName );
+	BiomeBuilder BiomeBuilder();
+	OreBuilder OreBuilder();
```
Comment (pau101) [inline src/main/java/com/mcmoddev/orespawn/api/os3/SpawnBuilder.java]: Uppercase method names, `newFeatureBuilder`, `newBiomeBuilder`, and `newOreBuilder` would be conventional.

File: src/main/java/com/mcmoddev/orespawn/impl/os3/BuilderLogicImpl.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		// for future expansion/orthagonality
+		return this;
+	}
+
+	@Override
+	public DimensionBuilder getDimension(String name) {
+		// I'd love for the following to be realistic...
+		// return this.DimensionBuilder(name);
+		Integer id = null;
+		switch(name.toLowerCase()) {
```
Comment (pau101) [inline src/main/java/com/mcmoddev/orespawn/impl/os3/BuilderLogicImpl.java]: Should be locale insensitive: `name.toLowerCase(Locale.ROOT)`

File: src/main/java/com/mcmoddev/orespawn/impl/os3/BuilderLogicImpl.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+public class BuilderLogicImpl implements BuilderLogic {
+	private final Map<Integer, DimensionBuilder> dimensions;
+	
+	public BuilderLogicImpl() {
+		this.dimensions = new HashMap<>();
+	}
+	
+	@Override
+	public DimensionBuilder DimensionBuilder(String name) {
+		switch(name.toLowerCase()) {
```
Comment (pau101) [inline src/main/java/com/mcmoddev/orespawn/impl/os3/BuilderLogicImpl.java]: Should be locale insensitive: `name.toLowerCase(Locale.ROOT)`

File: src/main/java/com/mcmoddev/orespawn/json/OS1Reader.java [hunk 7]
```diff
@@ -55,26 +60,28 @@ public static void loadEntries(Path confDir) {
...
 						// iterate over the dimensions
 		                for (JsonElement elem : root.get("dimensions").getAsJsonArray() ) {
 		                	JsonObject dim = elem.getAsJsonObject();
-		                	int theDim;
+		                	DimensionBuilder builder;
+		                	
 		                	if( "+".equals(dim.get("dimension").getAsString()) ) {
-		                		theDim = OreSpawnAPI.DIMENSION_WILDCARD;
+		                		builder = logic.DimensionBuilder();
+		                	} else if( dim.get("dimension").getAsString() == ""+dim.get("dimension").getAsInt() ) {
```
Comment (pau101) [inline src/main/java/com/mcmoddev/orespawn/json/OS1Reader.java]: String equality comparison with `==`, but more importantly this is not the way to check if the element is a number. The condition should be ` dim.get("dimension").isJsonPrimitive() && dim.get("dimension").getAsJsonPrimitive().isNumber()`, but `dim.get("dimension")` should be done once and stored in a variable before this if/else series.

File: src/main/java/com/mcmoddev/orespawn/api/plugin/PluginLoader.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        			Gson gson = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
+        			JsonObject json = gson.fromJson(reader, JsonObject.class);
+        			OS3Reader.loadFromJson(pd.modId, json);
+        		} catch (IOException e) {
+        			OreSpawn.LOGGER.error("Error creating a Buffered Reader to load Json from {} for mod {}",
+        					currentFile.toString(), pd.modId, e);
+        		} finally {
+        			if( reader != null) {
+        				try {
+							reader.close();
```
Comment (pau101) [inline src/main/java/com/mcmoddev/orespawn/api/plugin/PluginLoader.java]: This finally block can be replaced with `IOUtils.close(reader);`

File: src/main/java/com/mcmoddev/orespawn/impl/os3/OS3APIImpl.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    	for( Entry<String, BuilderLogic> ent : logic.entrySet() ) {
+    		for( Entry<Integer, DimensionBuilder> dL : ent.getValue().getAllDimensions().entrySet()) {
+				if( OreSpawn.spawns.containsKey(dL.getKey()) ) {
+					OreSpawn.spawns.get(dL.getKey()).addAll(dL.getValue().getAllSpawns());
+				} else {
+					OreSpawn.spawns.put(dL.getKey(), new ArrayList<>());
+					OreSpawn.spawns.get(dL.getKey()).addAll(dL.getValue().getAllSpawns());
+				}
+    		}
+        	OreSpawn.LOGGER.info("Registered spawn logic for mod " + ent.getKey());
```
Comment (pau101) [inline src/main/java/com/mcmoddev/orespawn/impl/os3/OS3APIImpl.java]: Change to: `OreSpawn.LOGGER.info("Registered spawn logic for mod {}", ent.getKey());`

File: src/main/java/com/mcmoddev/orespawn/api/plugin/IOreSpawnPlugin.java [hunk 10]
```diff
@@ -0,0 +1,8 @@
+package com.mcmoddev.orespawn.api.plugin;
+
+import com.mcmoddev.orespawn.api.os3.OS3API;
+
+@FunctionalInterface
```
Comment (pau101) [inline src/main/java/com/mcmoddev/orespawn/api/plugin/IOreSpawnPlugin.java]: @​FunctionalInterface isn't appropriate here.

**Review by pau101 — CHANGES_REQUESTED** [review]

**Comment by dshadowwolf** [comment]
latest patch added to this PR/review set makes sure the API is functional and disables the "load from a resource location" code as it is apparently bad/doesn't work - looks like the functionality for it was not patched in until Forge started working with the recipe stuff in 1.12

**Comment by dshadowwolf** [comment]
latest added patch-set adds a plugin to the code-base for registering vanilla ores (proving that it works) and fixes a lot of misc bugs.

**Comment by dshadowwolf** [comment]
And that patch, hopefully, gets resource-loading working
(quick testing shows that paths to resources in all jars on the classpath are accessible, yay!)

**Comment by dshadowwolf** [comment]
last couple patch sets have gotten the "plugins can have data in their jar files and we can copy them out and use them" issue handled - that code should be relatively stable.

**Comment by dshadowwolf** [comment]
Last commit added to this set removes all warnings from the Java side of things and all but 2 issues from SonarLint (so far) - SonarQube is likely to still have quite a bit after this is merged.

File: src/main/java/com/mcmoddev/orespawn/commands/AddOreCommand.java [hunk 11]
```diff
@@ -51,77 +55,59 @@ public void execute(MinecraftServer server, ICommandSender sender, String[] args
...
         } catch (NumberFormatException e) {
             throw new CommandException(args[1] + " isn't a valid dimension");
         }
 
         JsonObject ore = new JsonObject();
-        ore.addProperty("block", state.getBlock().getRegistryName().toString());
-        ore.addProperty("state", StateUtil.serializeState(state));
+        ore.addProperty(BLOCK, state.getBlock().getRegistryName().toString());
+        ore.addProperty(STATE2, StateUtil.serializeState(state));
         ore.addProperty("size", 25);
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/orespawn/commands/AddOreCommand.java]: We should probably extend addore so size et al can be specified as part of the command, or defaulted to as shown if not provided

File: src/main/java/com/mcmoddev/orespawn/impl/os3/BuilderLogicImpl.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+	private final Map<Integer, DimensionBuilder> dimensions;
+	
+	public BuilderLogicImpl() {
+		this.dimensions = new HashMap<>();
+	}
+	
+	@Override
+	public DimensionBuilder newDimensionBuilder(String name) {
+		switch(name.toLowerCase(Locale.ROOT)) {
+		case "overworld":
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/orespawn/impl/os3/BuilderLogicImpl.java]: make "overworld", "the nether" and "the end" shared constants, also provide "the overworld", "nether" and "end" for consistency.

File: src/main/java/com/mcmoddev/orespawn/json/OS1Reader.java [hunk 13]
```diff
@@ -55,26 +60,35 @@ public static void loadEntries(Path confDir) {
...
-		                		theDim = dim.get("dimension").getAsInt();
-		                	}
+		                	DimensionBuilder builder;
 		                	
-		                	DimensionLogic dimensionLogic = spawnLogic.getDimension(theDim);
-		                	
-		                	// now iterate the ores
-		                	for( JsonElement oElem : dim.get("ores").getAsJsonArray() ) {
-		                		loadOre( oElem.getAsJsonObject(), dimensionLogic );
+		                	JsonElement dimElem = dim.get("dimension");
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/orespawn/json/OS1Reader.java]: "dimension" and "ores" should really be shared constants.

File: src/main/java/com/mcmoddev/orespawn/json/OS1Reader.java [hunk 14]
```diff
@@ -84,45 +98,46 @@ public static void loadEntries(Path confDir) {
...
-            
-            for (JsonElement biomeEntry : biomesArray) {
-                Biome bm = ForgeRegistries.BIOMES.getValue(new ResourceLocation(biomeEntry.getAsString()));
+	private static void loadOres(JsonArray ores, DimensionBuilder builder) {
+		List<SpawnBuilder> spawns = new ArrayList<>();
+		for( JsonElement e : ores ) {
+			JsonObject ore = e.getAsJsonObject();
+			SpawnBuilder spawn = builder.newSpawnBuilder(null);
+			OreBuilder tO = spawn.newOreBuilder();
+			String blockID = ore.get("blockID").getAsString();
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/orespawn/json/OS1Reader.java]: "blockID" should be a shared constant.

File: src/main/java/com/mcmoddev/orespawn/json/OS1Reader.java [hunk 15]
```diff
@@ -84,45 +98,46 @@ public static void loadEntries(Path confDir) {
...
+			int meta = ore.has("metaData")?ore.get("metaData").getAsInt():0;
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/orespawn/json/OS1Reader.java]: "metaData" should be a shared constant.

File: src/main/java/com/mcmoddev/orespawn/json/OS1Reader.java [hunk 16]
```diff
@@ -84,45 +98,46 @@ public static void loadEntries(Path confDir) {
...
+			if( meta > 0 ) {
+				tO.setOre(blockID.toString(), meta);
+			} else {
+				tO.setOre(blockID.toString());
+			}
+			float frequency = ore.has("frequency")?ore.get("frequency").getAsFloat():20.0f;
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/orespawn/json/OS1Reader.java]: "frequency", "minHeight", "maxHeight", "variation", "default" and "size" should be shared constants

File: src/main/java/com/mcmoddev/orespawn/json/OS1Reader.java [hunk 17]
```diff
@@ -84,45 +98,46 @@ public static void loadEntries(Path confDir) {
...
 
-                if (bm != null) {
-                    biomes.add(bm);
-                }
-            }
-		} else {
-			biomes = Collections.<Biome>emptyList();
+			BiomeBuilder biomes = spawn.newBiomeBuilder();
+			if( ore.has("biomes") ) {
+				for( JsonElement bm : ore.get("biomes").getAsJsonArray() ) {
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/orespawn/json/OS1Reader.java]: "biomes" should be a shared constant

File: src/main/java/com/mcmoddev/orespawn/json/OS2Reader.java [hunk 18]
```diff
@@ -51,60 +48,60 @@ public static void loadEntries() {
...
-						SpawnLogic spawnLogic = OreSpawn.API.createSpawnLogic();
+						BuilderLogic logic = OreSpawn.API.getLogic(FilenameUtils.getBaseName(file.getName()));
+						List<DimensionBuilder> builders = new ArrayList<>();
 						
 		                for (JsonElement element : elements ) {
 		                    JsonObject object = element.getAsJsonObject();
 
-		                    int dimension = object.has("dimension") ? object.get("dimension").getAsInt() : OreSpawnAPI.DIMENSION_WILDCARD;
-		                    DimensionLogic dimensionLogic = spawnLogic.getDimension(dimension);
+		                    int dimension = object.has("dimension") ? object.get("dimension").getAsInt() : OreSpawn.API.dimensionWildcard();
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/orespawn/json/OS2Reader.java]: "dimension" should be a shared constant.

**Review by jriwanek — CHANGES_REQUESTED** [review]
I didn't tag *every* case where there's strings which should be constants for reuse throughout the code, but sharing them all is the intent.

**Review by Darkhax — COMMENTED** [review]
Mostly good, as Jas said using constants will help improve the maintainability of the code. I think the builder pattern may be a bit overboard here, and JavaDocs are in order, however I believe those are outside the scope for this review.

**Review by jriwanek — APPROVED** [review]

**Pull request merged**